model/DataObject.php

```diff
@@ -1876,164 +1932,256 @@ public function has_many($component = null, $classOnly = true) {
+		} else {
+			return false;
+		}
+
+		return ($classOnly) ? preg_replace('/(.+)?\..+/', '$1', $hasMany) : $hasMany;
+	}
+
+	/**
+	 * @deprecated 4.0 Method has been renamed to manyManyExtraFields()
+	 * @param string $component
+	 * @return array
+	 */
+	public function many_many_extraFields($component = null) {
+		Deprecation::notice('4.0', 'Method has been renamed to manyManyExtraFields()');
+		return $this->manyManyExtraFields($component);
+	}
+
+	/**
 	 * Return the many-to-many extra fields specification.
 	 *
 	 * If you don't specify a component name, it returns all
 	 * extra fields for all components available.
 	 *
 	 * @param string $component Name of component
-	 * @return array
+	 * @return array|null
 	 */
-	public function many_many_extraFields($component = null) {
-		$classes = ClassInfo::ancestry($this);
+	public function manyManyExtraFields($component = null) {
+		if($component) {
+			return $this->manyManyExtraFieldsForComponent($component);
+		}
 
+		$classes = ClassInfo::ancestry($this);
 		foreach($classes as $class) {
 			if(in_array($class, array('ViewableData', 'Object', 'DataObject'))) continue;
-			$relationName = null;
 
-			// Find extra fields for one component
-			if($component) {
-				$SNG_class = singleton($class);
-				$extraFields = $SNG_class->stat('many_many_extraFields');
+			// Find all the extra fields for all components
+			$newItems = eval("return (array){$class}::\$many_many_extraFields;");
 
-				// Extra fields are immediately available on this class
-				if(isset($extraFields[$component])) {
-					return $extraFields[$component];
+			foreach($newItems as $k => $v) {
+				if(!is_array($v)) {
+					user_error(
+						"$class::\$many_many_extraFields has a bad entry: "
+						. var_export($k, true) . " => " . var_export($v, true)
+						. ". Each many_many_extraFields entry should map to a field specification array.",
+						E_USER_ERROR
+					);
 				}
+			}
 
-				$manyMany = $SNG_class->stat('many_many');
-				$candidate = (isset($manyMany[$component])) ? $manyMany[$component] : null;
-				if($candidate) {
-					$SNG_candidate = singleton($candidate);
-					$candidateManyMany = $SNG_candidate->stat('belongs_many_many');
+			return $newItems;
+		}
+	}
 
-					// Find the relation given the class
-					if($candidateManyMany) foreach($candidateManyMany as $relation => $relatedClass) {
-						if($relatedClass == $class) {
-							$relationName = $relation;
-							break;
-						}
-					}
+	/**
+	 * Return the many-to-many extra fields specification for a specific component.
+	 * @param string $component
+	 * @return array|null
+	 */
+	public function manyManyExtraFieldsForComponent($component) {
+		$classes = ClassInfo::ancestry($this);
+		foreach($classes as $class) {
+			if(in_array($class, array('ViewableData', 'Object', 'DataObject'))) continue;
+			$relationName = null;
+			$SNG_class = singleton($class);
+			$extraFields = $SNG_class->stat('many_many_extraFields');
```

Comment:
What do you think of the name of this method, `manyManyExtraFieldsForComponent`?  I was torn between that and `manyManyComponentExtraFields`